connection_producer.go

```diff
@@ -0,0 +1,266 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package snowflake
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"database/sql"
+	"encoding/pem"
+	"fmt"
+	"github.com/hashicorp/errwrap"
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-secure-stdlib/parseutil"
+	"github.com/hashicorp/vault/sdk/database/helper/dbutil"
+	"net/url"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/database/helper/connutil"
+	"github.com/mitchellh/mapstructure"
+	"github.com/snowflakedb/gosnowflake"
+)
+
+var (
+	ErrInvalidSnowflakeURL           = fmt.Errorf("invalid connection URL format, expect <account_name>.snowflakecomputing.com/<db_name>")
+	ErrInvalidPrivateKey             = fmt.Errorf("failed to read provided private_key")
+	accountAndDBNameFromConnURLRegex = regexp.MustCompile(`^(.+)\.snowflakecomputing.com/(.+)$`) // Expected format: <account_name>.snowflakecomputing.com/<db_name>
+)
+
+type snowflakeConnectionProducer struct {
+	ConnectionURL            string      `json:"connection_url"`
+	MaxOpenConnections       int         `json:"max_open_connections"`
+	MaxIdleConnections       int         `json:"max_idle_connections"`
+	MaxConnectionLifetimeRaw interface{} `json:"max_connection_lifetime"`
+	Username                 string      `json:"username"`
+	Password                 string      `json:"password"`
+	PrivateKey               []byte      `json:"private_key"`
+	UsernameTemplate         string      `json:"username_template"`
+	DisableEscaping          bool        `json:"disable_escaping"`
+
+	Initialized           bool
+	RawConfig             map[string]any
+	Type                  string
+	maxConnectionLifetime time.Duration
+	logger                log.Logger
+	snowflakeDB           *sql.DB
+	mu                    sync.RWMutex
+}
+
+func (c *snowflakeConnectionProducer) secretValues() map[string]string {
+	return map[string]string{
+		c.Password: "[password]",
```

Comment:
Should we add PrivateKey here? I think that should prevent the key leaking in error messages?

Reply:
I don't think that will prevent leaks in the Connection() method though. Only implementations of the Database interface will get sanitized.